### PR TITLE
Implement MSC4069: Inhibit profile propagation

### PR DESCRIPTION
**Pull request by turt2live**
MSC: https://github.com/matrix-org/matrix-spec-proposals/pull/4069

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

**Comment by turt2live** [comment]
I'm not sure why sytest is failing here. Feels like a flake.

**Comment by clokep** [comment]
> I'm not sure why sytest is failing here. Feels like a flake.

https://github.com/matrix-org/sytest/issues/1201

**Review by erikjohnston — COMMENTED** [review]
This looks sane! I haven't done a full nit-picky review yet. Do you want this to land or are you happy having it as a PR as proof of implementation?

**Comment by turt2live** [comment]
It would be good to land: it's somewhat blocking implementation of MSCs in matrix-media-repo.

File: synapse/rest/client/profile.py [hunk 1]
```diff
@@ -27,6 +28,24 @@
...
+    if hs.config.experimental.msc4069_profile_inhibit_propagation:
+        do_propagate = request.args.get(b"org.matrix.msc4069.propagate")
+        if do_propagate is not None:
+            do_propagate = do_propagate[0].lower()
+            if do_propagate not in [b"true", b"false"]:
+                raise SynapseError(
+                    400, "'propagate' is malformed", errcode=Codes.INVALID_PARAM
+                )
+            else:
+                propagate = do_propagate == b"true"
```
Comment (erikjohnston) [inline synapse/rest/client/profile.py:45]:
```suggestion
        do_propagate = request.args.get(b"org.matrix.msc4069.propagate")
        if do_propagate is not None:
            propogate = parse_boolean(
                request, "org.matrix.msc4069.propagate", default=False
            )
```

I think should do it (with `from synapse.http.servlet import parse_boolean`)

**Review by erikjohnston — COMMENTED** [review]
Otherwise looks good

File: synapse/rest/client/profile.py (hunk 1 above)
Comment (turt2live) [reply synapse/rest/client/profile.py:45]: Applied manually

**Review by erikjohnston — APPROVED** [review]
Thanks!

**Pull request merged**